cmd/config-sync/collector.go

```diff
@@ -108,34 +108,41 @@ func runLeaderElection(lock *resourcelock.ConfigMapLock, ctx context.Context, id
 	)
 	begin := time.Now()
 	podname, _ := os.Hostname()
-	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
-		Lock:            lock,
-		ReleaseOnCancel: true,
-		LeaseDuration:   15 * time.Second,
-		RenewDeadline:   10 * time.Second,
-		RetryPeriod:     2 * time.Second,
-		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: func(c context.Context) {
-				log.Printf("COLLECTOR: Leader %s starting site collection after %s\n", podname, time.Since(begin))
-				leaderCtx, cancel = context.WithCancel(ctx)
-				siteCollector(leaderCtx, cli)
-				if err := startFlowController(leaderCtx, cli); err != nil {
-					log.Printf("COLLECTOR: Failed to start controller for emitting site events: %s", err)
-				}
+	for {
+		leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+			Lock:            lock,
+			ReleaseOnCancel: true,
+			LeaseDuration:   15 * time.Second,
+			RenewDeadline:   10 * time.Second,
+			RetryPeriod:     2 * time.Second,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(c context.Context) {
+					log.Printf("COLLECTOR: Leader %s starting site collection after %s\n", podname, time.Since(begin))
+					leaderCtx, cancel = context.WithCancel(ctx)
+					siteCollector(leaderCtx, cli)
+					if err := startFlowController(leaderCtx, cli); err != nil {
+						log.Printf("COLLECTOR: Failed to start controller for emitting site events: %s", err)
+					}
+				},
+				OnStoppedLeading: func() {
+					// No longer the leader, transition to inactive
+					cancel()
+					log.Printf("COLLECTOR: Lost leader after %s\n", time.Since(begin))
+				},
+				OnNewLeader: func(current_id string) {
+					if current_id == id {
+						// Remain as the leader
+						return
+					}
+					log.Printf("COLLECTOR: New leader for site collection is %s\n", current_id)
+				},
 			},
-			OnStoppedLeading: func() {
-				// No longer the leader, transition to inactive
-				cancel()
-			},
-			OnNewLeader: func(current_id string) {
-				if current_id == id {
-					// Remain as the leader
-					return
-				}
-				log.Printf("COLLECTOR: New leader for site collection is %s\n", current_id)
-			},
-		},
-	})
+		})
+		if ctx.Err() != nil {
```

Comment:
Do we need some kind of sleep or backoff?  If there are two pods competing for the lock, once one gets one, will the other just busy loop?

Reply:
That's probably a good call! In normal operation it shouldn't be strictly necessary, according to [docs](https://pkg.go.dev/k8s.io/client-go@v0.20.0/tools/leaderelection#RunOrDie) and a skim of the leaderelection code it looks like the only way we will ever hit this loop is when we are able to acquire the lock then fail to renew - so pods competing for the the lock should all be neatly polling at least until they can acquire the lock. I will throw in some limiting out of caution, though. It seems that by far the most popular usage of leaderelection is to call os.Exit from the `OnStoppedLeading` callback (which is not what I think we want), so we are in slightly uncharted waters here.

Reply:
I suppose I agree! In my observations when we lose this lock it is because it cannot read or update the lock (configmap) within that 10 second lease RenewDeadline, which seems exceptional (as we have been running this logic for years and just encountered this.)

Updated the PR to just log and exit. Want to do a little live testing before I merge.